### ISSUE TITLE
Encountered a truth rectangle located at...

### ISSUE THREAD

**Issue by kapraf** [davisking/dlib#320]
Hi Guys,
Didn't have a chance to thank you for this work. This is amazing.

Sorry for a rookie question but I struggle with something I don't understand in your example. Using imglab tool I defined my train/test object xmls but the dnn_mmod_ex keeps saying me:
"Encountered a truth rectangle located at [(10, 51) (23, 70)] that is too close to the edge
of the image to be captured by the CNN features."

Depending on the cropper and mmod_options settings the values in brackets change. It seens pretty self-explanatory comment so I wanted to know which exactly region causes the error. So I endded up with one image in the train xml that has one region in the center of it. Definitelly it is not too close to the edge of the image. I noticed that when the regions are too small this error can appear.

Could you please shed more light on this?

test-case included
[player-detection.tar.gz](https://github.com/davisking/dlib/files/575536/player-detection.tar.gz)

**Comment by AndreaPisoni** [comment]
Happens to me as well, I have even padded the image by 40% on every side so definitely there can't be anything close to the edges... 

Still get Encountered a truth rectangle located at [(0, 73) (34, 204)] etc... the coordinates are weird since there is no chance something is at 0... Maybe an issue with random_cropper?

**Comment by davisking** [comment]
That error comes from the trainer, which only sees the output of the random cropper.  So yes, it's definitely that the cropped images aren't what you expect.

**Comment by AndreaPisoni** [comment]
Any insights on how to debug that? I have not passed any weird parameter to the cropper... just the max_rotation_degrees...

**Comment by davisking** [comment]
Look at the images that come out with your eyes and see if it looks funny.

**Comment by AndreaPisoni** [comment]
I put the train step in a try-catch and showed the cropper windows when an exception is raised. 

The cropper windows seem normal, yes, there are (red) boxes that almost touch the edge in the crop, but I do not think I can do anything about this?

The other interesting thing is that I tried to ignore the error and just continue to the next train step. It is interesting because while the cropper keeps giving different crops at each step, the trainer raises the same exception on every step that is ran after the first exception is raised.

I.e. if the error is 

Encountered a truth rectangle located at [(0, 77) (27, 187)] that is too close to the edge
of the image to be captured by the CNN features.

Then it will return the same error at every subsequent iteration, with the same exact coordinates, even if the images are not the same.

**Comment by davisking** [comment]
So I looked at this.  If you pop up an image_window and look at each of the images produced by the random_cropper you will see that sometimes the player is very near the edge of the image.  You have to realize that the cropper is cropping out images from your training image.  They aren't all going to be centered on the players.  In particular, the random cropper will pick background images with some chance and if that happens to contain your guy he could appear anywhere.  The cropper will mark objects as ignored if they are partly outside the crop, which normally prevents you from getting this error.  However, since your player is so small the center of the player can get close to the edge of the image, and, due to the way the CNN downsamples an image some of the border pixels are lost, putting the guy's center of mass outside the image.  

Anyway, I don't really know what to do about this other than to tell users to be careful about the min-batches they create.  You should look at the images (use image_window and plot the boxes to see what is happening) and see if it looks like reasonable training data.  It should also be emphasized that you need to label all the players in the image or training will be messed up.  It will try to learn that all the other players aren't players.  Which brings me to another point, you might be thinking that you can disable the random background cropping to avoid this whole issue, but you would still have a problem of a crop that happens to contain multiple players in view, which obviously all can't appear in the center of the crop.

I could just remove that error message and make the training automatically ignore such boxes.  Which is fine from a machine learning point of view and would be a fine thing to do in your specific case.  However, I have a pretty strong conviction that doing so would lead to users creating really bad training sets that are getting silently ignored by the algorithm, which would lead to the algorithm not learning a sensible result, or anything at all since it might ignore all the data.  

As for what you can do, you could add a check inside the loop that generates the mini-batch for boxes that are very close to the edge of the image and mark them as ignore by setting mmod_rect::ignore to true.  This is easy to do with an if statement something like 
```
if (!get_rect(img).contains(grow_rect(some_player,10))) 
    some_player.ignore = true;
```

As an aside, you should probably use tighter boxes around the players and make them higher resolution.  When you look at the crops coming out of the random cropper you will see that they players are little more than small blurs, which is another reason why you should always look at the training data you are making.

**Comment by hexiangquan** [comment]
Encountered a truth rectangle located at [(13, 112) (28, 153)] that is too close to the edge
of the image to be captured by the CNN features. i do not how to address the problem?
